spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/Hive2IcebergAction.java

```diff
@@ -0,0 +1,445 @@
+import org.apache.iceberg.spark.source.SparkTable;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.catalog.CatalogTable;
+import org.apache.spark.sql.catalyst.catalog.CatalogUtils;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.StagingTableCatalog;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.catalog.TableChange;
+import org.apache.spark.sql.connector.catalog.V1Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Hive2IcebergAction extends BaseTableCreationSparkAction<Hive2IcebergAction>
+    implements Hive2Iceberg {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Hive2IcebergAction.class);
+  private SnapshotTableSparkAction snapshotTableSparkAction;
+  private Identifier sourceTableIdent;
+  private int parallelism = 1;
+  private SparkSession sparkSession;
+  private CatalogPlugin sourceCatalog;
+
+  Hive2IcebergAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableIdent) {
+    super(spark, sourceCatalog, sourceTableIdent);
+    this.sourceTableIdent = sourceTableIdent;
+    this.sparkSession = spark;
+    this.sourceCatalog = sourceCatalog;
+  }
+
+  @Override
+  public Hive2Iceberg parallelism(int pParallelism) {
+    this.parallelism = pParallelism;
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+
+    // Step1. Process SnapShot
+    String snapShotTableName = generateSnapShotTableName(sourceTableIdent.name());
+    boolean snapshotSuccess = handleSnapshotCreation(snapShotTableName);
+    if (!snapshotSuccess) {
+      return buildFailureResult("Snapshot creation failed");
+    }
+
+    // Step2. Prepare SourceTable
+    OperationSourceTableResult sourceTableResult = handleSourceTablePreparation();
+    if (!sourceTableResult.isSuccess()) {
+      return buildFailureResult("Source Table preparation failed");
+    }
+
+    CatalogTable sourceSparkTable = sourceTableResult.getSourceSparkTable();
+    Identifier sourceIdentifier = sourceTableResult.getSourceIdentifier();
+
+    // Step3. Prepare TargetTable
+    OperationTargetTableResult targetTableResult = handleTargetTablePreparation(snapShotTableName);
+    if (!targetTableResult.isSuccess()) {
+      return buildFailureResult("Target Table preparation failed");
+    }
+
+    SparkTable snapShotSparkTable = targetTableResult.getSnapShotSparkTable();
+
+    // Step4. Rewrite Table Path
+    OperationPathRewriteResult handlePathRewrite =
+        handlePathRewrite(snapShotSparkTable, sourceSparkTable);
+    if (!handlePathRewrite.isSuccess()) {
+      return buildFailureResult("Rewrite Table Path failed");
+    }
+
+    String latestVersion = handlePathRewrite.getLatestVersion();
+    String targetPrefix = CatalogUtils.URIToString(sourceSparkTable.storage().locationUri().get());
```

Comment:
Add null safety check
String targetPrefix = sourceSparkTable.storage().locationUri()
      .map(CatalogUtils::URIToString)
      .orElseThrow(() -> new IllegalStateException("Source table location URI is not available"));